File: SpherlyServer/src/Sphero.java
import java.io.IOException;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.Iterator;
import java.util.Vector;

import javax.bluetooth.RemoteDevice;

import org.java_websocket.WebSocket;


public class Sphero {
	private int id;
    private byte packet_number = 0;
    private boolean is_connected = false;
    private boolean enable_collision_detection = false;
    private SpheroHandler OnCollision = null;
    private boolean enable_power_notification = false;
    private SpheroHandler OnPowerNotification = null;
    private SpheroHandler OnConnectionErrorNotification = null;
    private SpherlyWebSocketServer server = null;
    private String address = "";
    private int test_error_limit = 0;
    private int test_error_counter = 0;

    //private onCollision
    private int heading = 0;
    
    public Thread connectionThread;

    public Sphero(SpherlyWebSocketServer server, int id) {
    	this.id = id;
        is_connected = false;
        enable_collision_detection = false;
        OnCollision = null;
        enable_power_notification = false;
        OnPowerNotification = null;
        heading = 0;
        this.server = server;
    }

    public static Vector<RemoteDevice> FindSpheros(boolean filter, SpherlyWebSocketServer server) {
        Vector<RemoteDevice> devices;
        try {
            devices = server.comm.findDevices();
            if (!filter) return devices;
            Vector<RemoteDevice> spheros = new Vector<RemoteDevice>();
            for (Iterator<RemoteDevice> i = devices.iterator(); i.hasNext(); ) {
                RemoteDevice device = i.next();
                if (device.getFriendlyName(false).contains("Sphero")) {
                    spheros.add(device);
                }
            }
            return spheros;
        } catch (Exception e) {
            StringWriter sw = new StringWriter();
            PrintWriter pw = new PrintWriter(sw);
            e.printStackTrace(pw);
            server.displayError(sw.toString());
            return null;
        }
    }

    public boolean connect(/*RemoteDevice device/**/String address/**/) {
        if (server.comm.connect(address)) {
        	this.address = address;
            is_connected = true;
            startListening();
            OnCollision = null;
            heading = 0;
            setHeading(heading);
            return true;
        } else {
            return false;
        }
    }
    
    public boolean reconnect(){
    	if (server.comm.connect(this.address)){
    		is_connected = true;
    		return true;
    	}else{
    		return false;
    	}
    }

    public void disconnect() {
        is_connected = false;
        server.webclientConnectedToSphero = false;
        server.comm.disconnect();
    }

    public boolean isConnected() {
        is_connected = server.comm.isConnected();
        return is_connected;
    }

    public void incPacketNum() {
        packet_number++;
        packet_number %= 256;
    }

    public void clearPacketNum() {
        packet_number = 0;
    }

    public byte[] buildCommand(byte device, byte command, byte[] data, boolean answer) {
        byte sop1 = (byte) 255;
        byte sop2 = (byte) 254;
        if (answer)
            sop2++;
        byte seq = packet_number;
        incPacketNum();
        byte dlen = (byte) (data.length + 1);
        byte dataSum = 0;
        for (int i = 0; i < data.length; i++) {
            dataSum += data[i];
        }
        byte chk = (byte) ((device + command + seq + dlen + dataSum) % 256);
        chk ^= 0xFF;
        byte packet[] = new byte[data.length + 7];
        packet[0] = sop1;
        packet[1] = sop2;
        packet[2] = device;
        packet[3] = command;
        packet[4] = seq;
        packet[5] = dlen;
        for (int i = 0; i < data.length; i++) {
            packet[6 + i] = data[i];
        }
        packet[packet.length - 1] = chk;
        return packet;
    }

    public void setHeading(int heading) {
    	this.heading = heading;
        byte device = 0x02;
        byte command = 0x01;
        byte lsbHeading = (byte) (heading % 256);
        byte msbHeading = (byte) (heading >> 8);
        byte[] data = new byte[]{msbHeading, lsbHeading};
        byte[] packet = buildCommand(device, command, data, false);
        server.comm.write(packet);
    }

    public void resetHeading(){
        heading = 0;
        setHeading(0);
    }

    public void setRGB(int red, int green, int blue) {
        byte device = 0x02;
        byte command = 0x20;
        byte flag = 0;
        byte[] data = new byte[]{(byte) red, (byte) green, (byte) blue, flag};
        byte[] packet = buildCommand(device, command, data, false);
        server.comm.write(packet);
    }

    public void rollForward(int speed) {
        rollForward(speed, 1);
    }

    public void rollForward(int speed, int state) {
        roll(heading, speed, state);
    }

    public void roll(int heading, int speed) {
        heading = (((heading % 360) + 360) % 360);
        roll(heading, speed, 1);
    }

    public void roll(int heading, int speed, int state) {
        byte device = 0x02;
        byte command = 0x30;
        byte lsbHeading = (byte) (heading % 256);
        byte msbHeading = (byte) (heading >> 8);
        byte[] data = new byte[]{
                (byte) speed, msbHeading, lsbHeading, (byte) state
        };
        byte[] packet = buildCommand(device, command, data, false);
        server.comm.write(packet);
    }

    public void turn(int direction) {
        heading += direction;
        heading = (((heading % 360) + 360) % 360);
        roll(heading, 0, 1);
    }

    public void setRotationRate(int rate) {
        byte device = 0x02;
        byte command = 0x03;
        byte[] data = new byte[]{(byte) rate};
        byte[] packet = buildCommand(device, command, data, false);
        server.comm.write(packet);
    }

    public void stop() {
        roll(heading, 0);
    }

    public void setStabilization(boolean flag) {
        byte device = 0x02;
        byte command = 0x02;

        int flag_data = 0;
        if (flag) {
            flag_data = 1;
        }

        byte[] data = new byte[]{(byte) flag_data};
        byte[] packet = buildCommand(device, command, data, false);
        server.comm.write(packet);
    }

    public void calibrate(boolean flagOn) {
        if (flagOn) {
            // enter in the calibration process
            setRGB(0, 0, 0);
            setBackLED(127);
            resetHeading();
            setStabilization(false);

        } else {
            // stop the calibration process
            setBackLED(0);
            setRGB(127, 127, 127);
            resetHeading();
            setStabilization(true);
        }
    }


    public void setInactivityTimeout(int time) {
        byte device = 0x00;
        byte command = 0x25;
        byte[] data = new byte[]{(byte) time};
        byte[] packet = buildCommand(device, command, data, false);
        System.out.println("setting inactivity timeout to: " + time + " seconds");
        server.comm.write(packet);
    }

    public void sleep() {
        byte device = 0x00;
        byte command = 0x22;
        byte[] data = new byte[]{0x00, 0x00, 0x00, 0x00, 0x00};
        byte[] packet = buildCommand(device, command, data, false);
        server.comm.write(packet);
    }

    public void setBackLED(int brightness) {
        byte device = 0x02;
        byte command = 0x21;
        byte[] data = new byte[]{(byte) brightness};
        byte[] packet = buildCommand(device, command, data, false);
        server.comm.write(packet);
    }

    //Dealing with responses from the sphero device
    public void enableCollisionDetection(boolean enable) {
        enable_collision_detection = enable;
        byte device = 0x02;
        byte command = 0x12;
        int enable_data = 0;
        if (enable) enable_data = 1;
        byte Xt = 0x32; //play with these values
        byte Yt = 0x32;
        byte Xspd = 0x64;
        byte Yspd = 0x64;
        int dead = 100;

        byte[] data = new byte[]{(byte) enable_data, Xt, Xspd, Yt, Yspd, (byte) dead};
        byte[] packet = buildCommand(device, command, data, false);
        server.comm.write(packet);
    }

    public void enablePowerNotification(boolean enable) {
        enable_power_notification = enable;
        byte device = 0x00;
        byte command = 0x21;
        int enable_data = 0;
        if (enable) enable_data = 1;
        byte[] data = new byte[]{(byte) enable_data};
        byte[] packet = buildCommand(device, command, data, false);
        server.comm.write(packet);
    }

    static class SpheroHandler {
    	WebSocket conn;
    	SpherlyWebSocketServer server;
    	public SpheroHandler(WebSocket conn, SpherlyWebSocketServer server){
    		this.conn = conn;
    		this.server = server;
    	}
        public void handle(){};
        public void handle(boolean data){};
        public void handle(byte data){};
    }

    public void setCollisionHandler(SpheroHandler handler) {
        OnCollision = handler;
    }

    public void setPowerNotificationHandler(SpheroHandler handler) {
        OnPowerNotification = handler;
    }
    
    public void setConnectionErrorHandler(SpheroHandler handler){
    	OnConnectionErrorNotification = handler;
    }

    interface SpheroListener extends Runnable {
        public byte[] readComm(int length) throws IOException;

        public void listen();
    }

    public void startListening() {
        final Sphero self = this;
        SpheroListener listener = new SpheroListener() {
            public byte[] readComm(int length) throws IOException{
                byte[] packet = server.comm.read(length);
                return packet;
            }

            public void listen() {
                displayMessage("listening to sphero...");
                boolean exceptionErrorOccurred = true;
                
                while (exceptionErrorOccurred){
	                while (self.isConnected()) {
	                    try {
	                    	//DEBUGGING CONNECTION ERRORS (stream closed)
	                    	if (test_error_counter < test_error_limit){
	                    		test_error_counter++;
	                    		throw new Exception("Jake Trower - testing stream closed");
	                    	}
	                    	
	                        byte[] header = readComm(5);
	                        /*StringBuilder sb = new StringBuilder();
						    for (byte b : header) {
						        sb.append(String.format("%02X ", b));
						    }
						    displayMessage(sb.toString());*/
	                        //sphero response packet
	                        //http://orbotixinc.github.io/Sphero-Docs/docs/sphero-api/packet-structures.html
	                        if (header == null) continue;
	
	                        byte sop1 = header[0];
	                        byte sop2 = header[1];
	
	                        //malformed packet
	                        if (sop1 != (byte) 0xff) {
	                            continue;
	                        }
	
	                        if (sop2 == (byte) 0xff) { //response
	                            int dlen = header[4];
	                            displayMessage("response. reading: " + dlen);
	                            readComm(dlen);
	                        } else if (sop2 == (byte) 0xfe) { //async packet
	                            byte idCode = header[2];
	
	                            if (idCode == (byte) 0x07) { //collision detection
	                                //displayMessage("Collision detected");
	                                if (self.enable_collision_detection && self.OnCollision != null) {
	                                    self.OnCollision.handle();
	                                }
	                            }
	
	                            int msbDlen = header[3] << 8;
	                            int lsbDlen = header[4];
	                            int dlen = msbDlen + lsbDlen;
	                            byte[] response = readComm(dlen);
	                            if (response == null) continue;
	                            if (idCode == (byte) 0x01) { //power notification!
	                                if (self.enable_power_notification) {
	                                    //displayMessage("Power notification: " + (int)response[0]);
	                                    self.OnPowerNotification.handle(response[0]);
	                                }
	                            }
	                            if (idCode == (byte) 0x07 && response.length >= 12) {
	                                //displayMessage("speed: "+(int)response[11]);
	                            }
	                        }
	                        exceptionErrorOccurred = false;
	                    } catch (Exception e) {
	                        StringWriter sw = new StringWriter();
	                        PrintWriter pw = new PrintWriter(sw);
	                        e.printStackTrace(pw);
	                        server.displayError(sw.toString());
	                        exceptionErrorOccurred = true;
	                    }
	                }
	                if (exceptionErrorOccurred){
	                    self.OnConnectionErrorNotification.handle(self.isConnected());
	                    exceptionErrorOccurred = false;
	                }
                }
          
                displayMessage("end listening...");
            }

            public void run() {
                listen();
            }
        };

        new Thread(listener).start();
    }

    public void displayMessage(String text) {
        server.displayMessage(text);
    }
}

File: __Linux/sphero.py
from pyRFCOMM import rfcomm
import thread

class Sphero:
	def __init__(self):
		self.comm = rfcomm.getRFCOMM()
		self.packetNumber = 0
		self.isConnected = False
		
		self.onCollision = None #callback to be used when a collision is detected
		self.heading = 0
	def findSpheros(self, filter=True):
		devices = self.comm.findDevices()
		if filter == False:
			return devices
		spheros = []
		for device in devices:
			if "Sphero" in device[1]:
				spheros.append(device)
		return spheros
	def connect(self, address):
		self.comm.connect(address)
		self.isConnected = True
		self.startListening()
		
		self.onCollision = None
		self.heading = 0
		self.setHeading(0)
	def disconnect(self):
		self.isConnected = False
		self.comm.disconnect()
	def isConnected(self):
		self.isConnected = self.comm.isConnected()
		return self.isConnected
	def incPacketNum(self):
		self.packetNumber += 1
		self.packetNumber %= 256
	def clearPacketNum(self):
		self.packetNumber = 0
	def buildCommand(self, device, command, data, answer=False): 
		#encodes the command and parameters into the protocol used by sphero
		sop1 = 255
		sop2 = 254
		if answer:
			sop2 += 1
		did = device
		cid = command
		seq = self.packetNumber
		self.incPacketNum()
		dlen = len(data) + 1
		data = data
		dataSum = 0
		for d in data:
			dataSum += d
		chk = (did + cid + seq + dlen + dataSum) % 256
		chk ^= 0xFF
		packet = bytearray([sop1, sop2, did, cid, seq, dlen])
		for d in data:
			packet.append(d)
		packet.append(chk)
		return packet
	def setRGB(self, red, green, blue):
		device = 0x02
		command = 0x20
		flag = 0
		data = bytearray([red, green, blue, flag])
		packet = self.buildCommand(device, command, data)
		self.comm.write(bytes(packet))
	def rollForward(self, speed, state=1):
		self.roll(self.heading, speed, state)
	def roll(self, heading, speed, state=1):
		print "heading:",heading," speed:",speed
		device = 0x02
		command = 0x30
		lsbHeading = heading % 256
		msbHeading = heading >> 8
		data = bytearray([speed, msbHeading, lsbHeading, state])
		packet = self.buildCommand(device, command, data)
		self.comm.write(bytes(packet))
	def turn(self, direction):
		self.heading += direction
		self.heading = self.heading % 360
		#self.setHeading(self.heading)
		self.roll(self.heading, 0, 1) #??? Does this work
	def setRotationRate(self, rate):
		print "set rotation:",rate
		device = 0x02
		command = 0x03
		data = bytearray([rate])
		packet = self.buildCommand(device, command, data)
		self.comm.write(bytes(packet))
	def setStabilization(self, flag):
		device = 0x02
		command = 0x02
		flag_byte = 0
		if (flag):
			flag_byte = 1
		data = bytearray([flag_byte])
		packet = self.buildCommand(device, command, data)
		self.comm.write(bytes(packet))
	def calibrate(self, flag):
		if (flag):
			self.setRGB(0, 0, 0)
			self.setBackLED(127)
			self.resetHeading()
			self.setStabilization(False)
		else:
			self.setBackLED(0)
			self.setRGB(127, 127, 127)
			self.resetHeading()
			self.setStabilization(True)
	def stop(self):
		self.roll(self.heading,0)
	def sleep(self):
		device = 0x00
		command = 0x22
		data = bytearray([0x00, 0x00, 0x00, 0x00, 0x00])
		packet = self.buildCommand(device, command, data)
		self.comm.write(bytes(packet))
	def resetHeading(self):
		self.setHeading(0)
	def setHeading(self, heading):
		self.heading = heading
		device = 0x02
		command = 0x01
		lsbHeading = heading % 256
		msbHeading = heading >> 8
		data = bytearray([msbHeading, lsbHeading])
		packet = self.buildCommand(device, command, data)
		self.comm.write(bytes(packet))
	def setBackLED(self, brightness):
		device = 0x02
		command = 0x21
		data = bytearray([brightness])
		packet = self.buildCommand(device, command, data)
		self.comm.write(bytes(packet))
	def setCollisionHandler(self, function):
		self.onCollision = function
	def enableCollisionDetection(self, enable=True):
		device = 0x02
		command = 0x12
		if enable:
			method = 0x01
		else:
			method = 0x00
		Xt = 0x56 #play with these values
		Yt = 0x56 #this one too
		Xspd = 0x64
		Yspd = 0x64
		dead = 100
		data = bytearray([method, Xt, Xspd, Yt, Yspd, dead])
		packet = self.buildCommand(device, command, data)
		self.comm.write(bytes(packet))
	def startListening(self):
		def readComm(length):
			#ideally, self.com.read would guarantee a read length that was given to it, but it doesn't.
			packet = self.comm.read(length)
			while (len(packet) < length):
				packet += (self.comm.read(length - len(packet)))
			return packet
	
		#start a new thread to listen for packets that indicate a collision
		def listen():
			print "begin listening\n"
			fullheader = []
			while self.isConnected:
				try:
					header = readComm(5)
					
					print "got a packet!, length:", len(header)
					
					#sphero response packet: http://orbotixinc.github.io/Sphero-Docs/docs/sphero-api/packet-structures.html
					#OKAY NOW LET'S PROCESS THE PACKET
					
					if ord(header[1]) == 0xff: #response packet
						dlen = ord(header[4])
						print "response. reading:", dlen
						readComm(dlen)
					elif ord(header[1]) == 0xfe: #async packet
						print "async"
						idCode = ord(header[2])
						if idCode == 0x07: #collision detection
							print "collision detected"
							if self.onCollision != None:
								self.onCollision()
						msbDlen = ord(header[3]) << 8
						lsbDlen = ord(header[4])
						dlen = msbDlen + lsbDlen
						print "reading:",dlen
						readComm(dlen)
				except Exception, e: 
					print "error:",str(e) #what is error return without exception set?
					pass
			print "end listening...\n"
		thread.start_new_thread(listen, ())
